### PR TITLE
Share a single MarkdownPipeline

### PR DESCRIPTION
**Pull request by yufeih**
Time to build _azure-docs-pr_ without _loading commits_, _run legacy_, _copy resource_ drops to 2 minutes.

#2972

### CODE REVIEW

File: src/docfx/lib/markdown/ExtractTitle.cs [hunk 1]
```diff
@@ -13,32 +12,27 @@ namespace Microsoft.Docs.Build
...
         private static string RenderTitle(HeadingBlock h1)
         {
-            using (var writer = new StringWriter())
-            {
-                var renderer = new HtmlRenderer(writer);
-                var pipeline = new MarkdownPipelineBuilder().Build();
-                pipeline.Setup(renderer);
-                renderer.Render(h1);
-                writer.Flush();
+            var writer = new StringWriter();
```
Comment (herohua) [inline src/docfx/lib/markdown/ExtractTitle.cs]: >StringWriter [](start = 29, length = 12)

dispose?

File: src/docfx/build/markdown/Markup.cs [hunk 2]
```diff
@@ -32,91 +27,101 @@ internal static class Markup
             { "Caution", "<p>Caution</p>" },
         };
 
+        private static readonly MarkdownPipeline s_markdownPipeline = CreateMarkdownPipeline();
+
+        [ThreadStatic]
+        private static (MarkupResult result, DependencyMapBuilder dependencyMap, Action<Document> buildChild) t_context;
```
Comment (herohua) [inline src/docfx/build/markdown/Markup.cs]: >t_context [](start = 110, length = 9)

do we have a better name? `t_context` and `Context` look confusing to me.

File: src/docfx/build/markdown/Markup.cs [hunk 3]
```diff
@@ -32,91 +27,101 @@ internal static class Markup
...
+
+        public static ref MarkupResult Context => ref t_context.result;
```
Comment (live1206) [inline src/docfx/build/markdown/Markup.cs]: >ref [](start = 22, length = 3)

just out of curiosity, why do we need "ref" here?

File: src/docfx/build/markdown/Markup.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/markdown/Markup.cs]: `t_` is the prefix for thread static fields as described [here](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md#c-coding-style)

File: src/docfx/lib/markdown/ExtractTitle.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/markdown/ExtractTitle.cs]: StringWriter dispose is an no-op

---
In reply to: [201658646](https://github.com/dotnet/docfx/pull/3043#discussion_r201658646) [](ancestors = 201658646)

**Pull request merged**